### PR TITLE
windows_function_app_slot fix slot name formatting #16305

### LINKED ISSUE #15784: azurerm_linux_function_app_slot slot name
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request

### Terraform (and AzureRM Provider) Version

* Terraform Core version: 1.1.7
* AzureRM Provider version: 2.98.0

### Terraform Configuration Files

```hcl
provider "azurerm" {
  features {}
}

resource "azurerm_resource_group" "example" {
  name     = "example-resources"
  location = "West Europe"
}

resource "azurerm_storage_account" "example" {
  name                     = "linuxfunctionappsa"
  resource_group_name      = azurerm_resource_group.example.name
  location                 = azurerm_resource_group.example.location
  account_tier             = "Standard"
  account_replication_type = "LRS"
}

resource "azurerm_service_plan" "example" {
  name                = "example-app-service-plan"
  resource_group_name = azurerm_resource_group.example.name
  location            = azurerm_resource_group.example.location
  os_type             = "Linux"
  sku_name            = "Y1"
}

resource "azurerm_linux_function_app" "example" {
  name                 = "example-linux-function-app"
  resource_group_name  = azurerm_resource_group.example.name
  location             = azurerm_resource_group.example.location
  service_plan_id      = azurerm_service_plan.example.id
  storage_account_name = azurerm_storage_account.example.name

  site_config {}
}

resource "azurerm_linux_function_app_slot" "example" {
  name                 = "example-linux-function-app-slot"
  function_app_id      = azurerm_linux_function_app.example.id
  storage_account_name = azurerm_storage_account.example.name

  site_config {}
}
```

### Description / Feedback

The `name` of the slot, as defined above, will result in an actual slot-name `example-linux-function-app-example-linux-function-app-slot` on Azure.
To achieve the intended slot name `example-linux-function-app-slot`, one would have to define the slot as follows:

```patch
 resource "azurerm_linux_function_app_slot" "example" {
-   name                 = "example-linux-function-app-slot"
+   name                 = "slot"
   function_app_id      = azurerm_linux_function_app.example.id
  storage_account_name = azurerm_storage_account.example.name

   site_config {}
 }
```

Terraform will most likely not create this resource, as it checks for the uniqueness of the slot-name, i.e. `slot`, instead of the combined slot-and-app-name `example-linux-function-app-slot`, like it will be on Azure:

https://github.com/hashicorp/terraform-provider-azurerm/blob/7ab536cfa67a2f7a3401445d97c799033a4e513c/internal/services/appservice/linux_function_app_slot_resource.go#L329-L332

Example above is taken [from the docs](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/linux_function_app_slot). A bug fix will have to adjust it as well.

### References

[From the Azure REST documentation](https://docs.microsoft.com/en-us/rest/api/appservice/web-apps/create-instance-function-slot) we can see that in order to create a slot, only the _slot name_ must be passed. Not the resulting full name.

### LINKED ISSUE #16305: azurerm_windows_function_app_slot now requires a globally unique name for the slot itself
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

0.15.4

### AzureRM Provider Version

3.0.2

### Affected Resource(s)/Data Source(s)

azurerm_windows_function_app_slot

### Terraform Configuration Files

```hcl
resource "azurerm_windows_function_app" "function_app" {
  service_plan_id             = var.asp_id
  location                    = var.region
  name                        = testapp-azf"
  resource_group_name         = var.resource_group_name
  storage_account_access_key  = var.sa_primary_access_key
  storage_account_name        = var.sa_name
  functions_extension_version = var.fa_extensions_version
  builtin_logging_enabled     = var.fa_builtin_logging_enabled

resource "azurerm_windows_function_app_slot" "function_app_slot" {
  name                        = "dev"
  function_app_id             = var.fa_id
  storage_account_access_key  = var.sa_primary_access_key
  storage_account_name        = var.sa_name
  functions_extension_version = var.fa_extensions_version
  builtin_logging_enabled     = var.fa_builtin_logging_enabled
```

### Debug Output/Panic Output

```shell
│ Error: the Site Name "testapp-azf" failed the availability check: Hostname 'dev' already exists. Please select a different name.
│ 
│   with module.fa_slot.azurerm_windows_function_app_slot.function_app_slot,
│   on .terraform/modules/fa_slot/main.tf line 1, in resource "azurerm_windows_function_app_slot" "function_app_slot":
│    1: resource "azurerm_windows_function_app_slot" "function_app_slot" {
│ 
│ the Site Name "testapp-azf" failed the availability check:
│ Hostname 'dev' already exists. Please select a different name.
```

### Expected Behaviour

The slot would be created with the Function app name and the slot name appended with a forward slash and would be globally  unique.
Example:
testapp-azf/dev

### Actual Behaviour

Error is thrown: Hostname 'dev' already exists.
The slot will only deploy if it is given its own globally unique name.

### Steps to Reproduce

Run a plan with the slot named something common that wouldn't be globally unique on its own.

### Important Factoids

This works as expected with the new resource: azurerm_windows_web_app_slot and previously worked correctly with  the azurerm_function_app_slot resource 

### References

_No response_

### PR DESCRIPTION
**Pull request by AaldertVanWijk-TomTom**
The current check for slot names in windows_function_app_slot and linux_function_app_slot is wrong.
It does not check for the full slot name but only for the extension part. 

This means it stops the deployment for names that don't necessarily exist but just happen to be common. And it will pass the check for names that do already exist.

Fixes #16305 
FIxes #15784

### CODE REVIEW

**Comment by AaldertVanWijk-TomTom** [comment]
Obviously I tested this in my own subscription, the API that is used to check the offending name can also easily be tested here:
https://docs.microsoft.com/en-us/rest/api/appservice/check-name-availability/check-name-availability#code-try-0

**Review by tiagoboeing — APPROVED** [review]

**Comment by tiagoboeing** [comment]
@katbyte could you direct this fix?

**Comment by dangerousCoderC** [comment]
Why is this taking so long??? I can say that quite a lot of people are waiting on this fix, crucial for proper slot usage...

**Comment by tiagoboeing** [comment]
Anyone? @tombuildsstuff @mbfrahry

**Comment by wc-whiteheadd** [comment]
@katbyte @jackofallops 

Hey guys,

can this please be looked at and prioritised, we have now been 2+ months with no staging slot due to such a small issue, even when its already fixed and just needs merging in correctly.

fyi i have added you guys into this thread as you have both just been on the latest merge around this area but for the zip deployments work.

**Comment by jackofallops** [comment]
Apologies for the delays here - this somehow slipped my notifications. I'm going to push a couple minor changes and run this through the test suite, we should be able to merge for this week's release.

**Comment by wc-whiteheadd** [comment]
@jackofallops thanks for this, its greatly appreciated.

File: internal/services/appservice/linux_function_app_slot_resource.go [hunk 1]
```diff
@@ -344,7 +344,7 @@ func (r LinuxFunctionAppSlotResource) Create() sdk.ResourceFunc {
 			}
 
 			availabilityRequest := web.ResourceNameAvailabilityRequest{
-				Name: utils.String(functionAppSlot.Name),
+				Name: utils.String(fmt.Sprintf("%s-%s", functionAppId.SiteName, functionAppSlot.Name)),
```
Comment (jackofallops) [inline internal/services/appservice/linux_function_app_slot_resource.go:347]: We should use the values from the ID here:
```suggestion
				Name: utils.String(fmt.Sprintf("%s-%s", id.SiteName, id.SlotName)),
```

File: internal/services/appservice/windows_function_app_slot_resource.go [hunk 2]
```diff
@@ -350,7 +350,7 @@ func (r WindowsFunctionAppSlotResource) Create() sdk.ResourceFunc {
 			}
 
 			availabilityRequest := web.ResourceNameAvailabilityRequest{
-				Name: utils.String(functionAppSlot.Name),
+				Name: utils.String(fmt.Sprintf("%s-%s", functionAppId.SiteName, functionAppSlot.Name)),
```
Comment (jackofallops) [inline internal/services/appservice/windows_function_app_slot_resource.go:353]: We should use the values from the ID here:
```suggestion
				Name: utils.String(fmt.Sprintf("%s-%s", id.SiteName, id.SlotName)),
```

**Review by jackofallops — APPROVED** [review]
Thanks @AaldertVanWijk-TomTom - Apologies for the delayed review here. I'll apply the review suggestions and get this into test so we can get it merged asap.

**Comment by jackofallops** [comment]
Tests are looking good:

image

**Pull request merged**